Removed use of iteration through a HashMap/HashSet in rustc_incremental and replaced with IndexMap/IndexSet

ndrewxie opened the pull request:
This allows for the `#[allow(rustc::potential_query_instability)]` in rustc_incremental to be removed, moving towards fixing #84447 (although a LOT more modules have to be changed to fully resolve it). Only HashMaps/HashSets that are being iterated through have been modified (although many structs and traits outside of rustc_incremental had to be modified as well, as they had fields/methods that involved a HashMap/HashSet that would be iterated through)

I'm making a PR for just 1 module changed to test for performance regressions and such, for future changes I'll either edit this PR to reflect additional modules being converted, or batch multiple modules of changes together and make a PR for each group of modules.

ndrewxie committed:
Removed use of iteration through a HashMap/HashSet in rustc_incremental and replaced with IndexMap/IndexSet

ndrewxie commented:
Aah it appears some tests are failing, ig I didn't hit all the test cases on my local machine while testing. I'll make a few edits, then retry

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4636067458/jobs/8203630025) [(plain)](https://github.com/rust-lang/rust/commit/a9c089c5270c7ec8bd17c1aad797d53e1707e7da/checks/12586275083/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking cranelift-frontend v0.93.0
    Checking cranelift-native v0.93.0
    Checking cranelift-object v0.93.0
    Checking rustc_codegen_cranelift v0.1.0 (/checkout/compiler/rustc_codegen_cranelift)
error[E0053]: method `join_codegen` has an incompatible type for trait
    |
    |
226 |     ) -> Result<(CodegenResults, FxHashMap<WorkProductId, WorkProduct>), ErrorGuaranteed> {
    |          |
    |          |
    |          expected `IndexMap<WorkProductId, ..., ...>`, found `HashMap<WorkProductId, ..., ...>`
    |          help: change the output type to match the trait: `std::result::Result<(CodegenResults, indexmap::map::IndexMap<WorkProductId, WorkProduct, BuildHasherDefault<FxHasher>>), ErrorGuaranteed>`
    |
    = note: expected signature `fn(&CraneliftCodegenBackend, std::boxed::Box<_>, &Session, &OutputFilenames) -> std::result::Result<(CodegenResults, indexmap::map::IndexMap<WorkProductId, WorkProduct, BuildHasherDefault<FxHasher>>), _>`
               found signature `fn(&CraneliftCodegenBackend, std::boxed::Box<_>, &Session, &OutputFilenames) -> std::result::Result<(CodegenResults, HashMap<WorkProductId, WorkProduct, BuildHasherDefault<FxHasher>>), _>`
For more information about this error, try `rustc --explain E0053`.
error: could not compile `rustc_codegen_cranelift` due to previous error
Build completed unsuccessfully in 0:01:40
```

</details>

cuviper commented:
Note that `IndexMap` iterator order is only deterministic with respect to the insertion order, so you must also ensure that all insertions and removals are also deterministic to get the kind of stability that rustc wants. Or you can take steps like sorting the map after that's done, so iteration will follow the sort order.

(I'm not saying this PR is necessarily a problem in that way, just noting it for caution.)

ndrewxie committed:
Updated cranelift codegen to reflect modified trait signature

ndrewxie commented:
Aah, I see. After I fix the issue with cranelift build failing, I'll go investigate that, and keep it in mind for future changes as well :)

ndrewxie commented:
Also just a quick question - I failed the build test for codegen_cranelift and codegen_gcc because the tests I was doing on my local machine never hit those modules. Is there a way I can test those as well? 

For codegen_cranelift, I found using --target=i686-pc-windows-gnu worked, but is there an equivalent target I can use to make sure my code doesn't break codegen_gcc?

Thanks!

bjorn3 commented:
`./x.py check` should check both cg_clif and cg_gcc. `./x.py build` will only build them if they are listed in the `codegen-backends` option of `config.toml`.

ndrewxie commented:
> `./x.py check` should check both cg_clif and cg_gcc. `./x.py build` will only build them if they are listed in the `codegen-backends` option of `config.toml`.

ohhhh I see, I haven't really been using ./x.py check... maybe I should start doing so, my ~~tabletop jet engine~~ laptop would certainly thank me 🤣

ndrewxie committed:
Fixed compiler error

Nilstrieb commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 916c1ca501f3e134b8cd8e61201d5a01a9b19d12 with merge e709771db4bde2ba3203fc1eb0b315b71b35c8c6...
<!-- homu: {"type":"TryBuildStarted","head_sha":"916c1ca501f3e134b8cd8e61201d5a01a9b19d12","merge_sha":"e709771db4bde2ba3203fc1eb0b315b71b35c8c6"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4641807584/jobs/8216004288)
Build commit: e709771db4bde2ba3203fc1eb0b315b71b35c8c6 (`e709771db4bde2ba3203fc1eb0b315b71b35c8c6`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4641807584/jobs/8216004288"},"merge_sha":"e709771db4bde2ba3203fc1eb0b315b71b35c8c6"} -->

rust-timer commented:
Queued e709771db4bde2ba3203fc1eb0b315b71b35c8c6 with parent da636956dd13638a5582f9dcd4699559d85ac2d5, future [comparison URL](https://perf.rust-lang.org/compare.html?start=da636956dd13638a5582f9dcd4699559d85ac2d5&end=e709771db4bde2ba3203fc1eb0b315b71b35c8c6).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (e709771db4bde2ba3203fc1eb0b315b71b35c8c6): [comparison URL](https://perf.rust-lang.org/compare.html?start=da636956dd13638a5582f9dcd4699559d85ac2d5&end=e709771db4bde2ba3203fc1eb0b315b71b35c8c6&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=da636956dd13638a5582f9dcd4699559d85ac2d5&end=e709771db4bde2ba3203fc1eb0b315b71b35c8c6&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=da636956dd13638a5582f9dcd4699559d85ac2d5&end=e709771db4bde2ba3203fc1eb0b315b71b35c8c6&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.7% | [2.5%, 2.9%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=da636956dd13638a5582f9dcd4699559d85ac2d5&end=e709771db4bde2ba3203fc1eb0b315b71b35c8c6&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

lcnr commented:
with https://github.com/rust-lang/compiler-team/issues/533 we should be able to instead use https://doc.rust-lang.org/nightly/nightly-rustc/rustc_data_structures/unord/struct.UnordMap.html for a lot of these i think? :thinking:

not sure what's the current status here but it feels preferable to use these whenever possible, cc @michaelwoerister

ndrewxie commented:
> with https://github.com/rust-lang/compiler-team/issues/533 we should be able to instead use https://doc.rust-lang.org/nightly/nightly-rustc/rustc_data_structures/unord/struct.UnordMap.html for a lot of these i think? 🤔

Ooo, that's a pretty cool data structure, didn't know that existed! Yeah, I guess that provides an even better determinism guarantee than FxIndexMap, because we either have to work with it in an order-agnostic way, or sort it. IDK if it'll cause performance regressions from extra .sorts() (in places where iteration by Fn instead of FnMut isn't sufficient), but I'll be happy to migrate over (whenever possible) and run those performance tests if michaelwoerister confirms that it's indeed the best practice :)

michaelwoerister commented:
IndexMap and UnordMap can both make things deterministic but have slightly different properties:

- IndexMap makes sure that iteration order matches insertion order. That means that one has to be careful to insert things in a deterministic way. E.g. the result of `some_regular_hashmap.iter().collect::<IndexMap<_, _>>()` will still not be deterministic, and it's not always easy to see if insertion order is deterministic.
- UnordMap "erases" the iteration order. It is safe to feed things into it in any order one wants. It's API allows to do iteration-order-independent operations because they are safe to do (and often that's all that's needed). It also allows to extract its items into an ordered collection, but only in a deterministic way. However, this can be expensive operation since we have to sort the items in a way that is stable across compilation sessions.

I've been trying to use UnordMap where possible because it is a bit more robust (i.e. it introduces a kind of indeterminism firewall). But for maps where it's trivially obvious that they are built deterministically, IndexMap is a good choice too.

lcnr commented:
@ndrewxie do you want to try to move some of these uses to `UnordMap`? please retag this to `S-waiting-to-review` whenever you're happy. I think it's fine to use `IndexMap` in places where it's not stricty necessary but would prefer to replace as much as possible with `UnordMap`

@rustbot author

ndrewxie commented:
> 

Alright, will do 👍

ndrewxie committed:
Switched some uses to UnordMap

ndrewxie committed:
Sorted a FxIndexSet for consistent iteration order

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4910048589/jobs/8766853556) [(plain)](https://github.com/rust-lang/rust/commit/10ba8b563901b0fcf66c27c6404c53353b6c878a/checks/13299344760/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

failures:

---- persist::fs::tests::test_all_except_most_recent stdout ----
thread 'persist::fs::tests::test_all_except_most_recent' panicked at 'assertion failed: `(left == right)`
  left: `UnordSet { inner: {"4", "1", "5", "2", "3"} }`,
 right: `UnordSet { inner: {"4", "1", "2", "3"} }`', compiler/rustc_incremental/src/persist/fs/tests.rs:14:5


failures:
    persist::fs::tests::test_all_except_most_recent
    persist::fs::tests::test_all_except_most_recent

test result: FAILED. 2 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 671.11µs

error: test failed, to rerun pass `-p rustc_incremental --lib`
```

</details>

ndrewxie committed:
Fixed failing test + minor cleanup

jyn514 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 2a85b83594f14126614e95a8dcbe17858c85dfdd with merge df9b0984ab3b60ab2aab2536f97ea071eb2a7510...
<!-- homu: {"type":"TryBuildStarted","head_sha":"2a85b83594f14126614e95a8dcbe17858c85dfdd","merge_sha":"df9b0984ab3b60ab2aab2536f97ea071eb2a7510"} -->

ndrewxie commented:
@rustbot review

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4910553218/jobs/8768906098)
Build commit: df9b0984ab3b60ab2aab2536f97ea071eb2a7510 (`df9b0984ab3b60ab2aab2536f97ea071eb2a7510`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4910553218/jobs/8768906098"},"merge_sha":"df9b0984ab3b60ab2aab2536f97ea071eb2a7510"} -->

rust-timer commented:
Queued df9b0984ab3b60ab2aab2536f97ea071eb2a7510 with parent ad6b20bf5225c65aab30169c9a173719f5bdfc04, future [comparison URL](https://perf.rust-lang.org/compare.html?start=ad6b20bf5225c65aab30169c9a173719f5bdfc04&end=df9b0984ab3b60ab2aab2536f97ea071eb2a7510).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (df9b0984ab3b60ab2aab2536f97ea071eb2a7510): [comparison URL](https://perf.rust-lang.org/compare.html?start=ad6b20bf5225c65aab30169c9a173719f5bdfc04&end=df9b0984ab3b60ab2aab2536f97ea071eb2a7510&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=ad6b20bf5225c65aab30169c9a173719f5bdfc04&end=df9b0984ab3b60ab2aab2536f97ea071eb2a7510&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ad6b20bf5225c65aab30169c9a173719f5bdfc04&end=df9b0984ab3b60ab2aab2536f97ea071eb2a7510&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.1% | [-2.1%, -2.1%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.1% | [-2.1%, -2.1%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ad6b20bf5225c65aab30169c9a173719f5bdfc04&end=df9b0984ab3b60ab2aab2536f97ea071eb2a7510&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.9%  | [2.9%, 2.9%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -5.4% | [-6.1%, -4.6%] | 5     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=ad6b20bf5225c65aab30169c9a173719f5bdfc04&end=df9b0984ab3b60ab2aab2536f97ea071eb2a7510&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 656.342s -> 656.729s (0.06%)

ndrewxie commented:
Aaah I was just thinking and realized that my use of `for_each` with a Fn in UnordMap/UnordSet kinda defeats the purpose of the collections (even with only Fn, not FnMut), because it still causes indeterminism in IO, as well as allowing the use of RefCells to break out of the "can't do anything while iterating that is order sensitive".

It's late for me so I'll fix that tomorrow, for tonight (this morning?) I'll just do a 
@rustbot author

ndrewxie committed:
Fixed unord mistake

workingjubilee commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 69ba7dc664e7a7294fdaff7655e3f66013f1cd3d with merge bd30887d9d604b103a42537022a18ae9352a73e9...
<!-- homu: {"type":"TryBuildStarted","head_sha":"69ba7dc664e7a7294fdaff7655e3f66013f1cd3d","merge_sha":"bd30887d9d604b103a42537022a18ae9352a73e9"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4920118574/jobs/8790208559)
Build commit: bd30887d9d604b103a42537022a18ae9352a73e9 (`bd30887d9d604b103a42537022a18ae9352a73e9`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4920118574/jobs/8790208559"},"merge_sha":"bd30887d9d604b103a42537022a18ae9352a73e9"} -->

rust-timer commented:
Queued bd30887d9d604b103a42537022a18ae9352a73e9 with parent dfe31889e10e36eed53327d1ca624fbf21b475a5, future [comparison URL](https://perf.rust-lang.org/compare.html?start=dfe31889e10e36eed53327d1ca624fbf21b475a5&end=bd30887d9d604b103a42537022a18ae9352a73e9).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (bd30887d9d604b103a42537022a18ae9352a73e9): [comparison URL](https://perf.rust-lang.org/compare.html?start=dfe31889e10e36eed53327d1ca624fbf21b475a5&end=bd30887d9d604b103a42537022a18ae9352a73e9&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=dfe31889e10e36eed53327d1ca624fbf21b475a5&end=bd30887d9d604b103a42537022a18ae9352a73e9&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.4% | [-0.5%, -0.3%] | 4     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=dfe31889e10e36eed53327d1ca624fbf21b475a5&end=bd30887d9d604b103a42537022a18ae9352a73e9&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.7% | [2.7%, 2.7%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=dfe31889e10e36eed53327d1ca624fbf21b475a5&end=bd30887d9d604b103a42537022a18ae9352a73e9&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=dfe31889e10e36eed53327d1ca624fbf21b475a5&end=bd30887d9d604b103a42537022a18ae9352a73e9&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 658.118s -> 657.89s (-0.03%)

ndrewxie commented:
@rustbot review

lcnr commented on compiler/rustc_incremental/src/assert_module_sources.rs:
```diff
@@ -118,9 +118,8 @@ impl<'tcx> AssertModuleSource<'tcx> {
-                self.available_cgus.iter().map(|cgu| cgu.as_str()).collect();
-            cgu_names.sort();
+            let cgu_names: Vec<String> =
+                self.available_cgus.items().map(|cgu| cgu.as_str().to_owned()).into_sorted(&());
```
i feel like this impl should be extendedd to all types whicih implement `HashStable` https://doc.rust-lang.org/nightly/nightly-rustc/rustc_data_structures/stable_hasher/trait.ToStableHashKey.html#impl-ToStableHashKey%3CHCX%3E-for-String

with that we should be able to keep using a `Vec<&str>` here? 🤔

lcnr replied:
also, please use `to_sorted_stable_ord` for every type which implements `StableOrd` itself. `&str` should implement `StableOrd`

ndrewxie replied:
Whoops, forgot this PR existed! My bad

I tried implementing ToStableHashKey for all HashStable implementers using a `impl<T, HCX> ToStableHashKey<HCX> for T where T: HashStable<HCX>`, but that runs into issues where it conflicts with `HashStable`s that already have an explicit ToStableHashKey implementation. Should I still look into doing this (probably could make a macro to implement this for a bunch of types), or just leave it as is?

As for the `to_sorted_stable_ord`, done 👍. There's a few types that appeared to be missing StableOrd impls, most notably `&str` and the tuples (provided that the elements within the tuple are StableOrd themselves). I went ahead and just added the impl for these types - is that correct?

michaelwoerister replied:
`ToStableHashKey` (especially when implementing it via `HashStable`) is a pretty heavy hammer and I've personally never been very happy about the existence of the `ToStableHashKey` to begin with. So my recommendation would be to not invest into it too much.

> There's a few types that appeared to be missing StableOrd impls, most notably &str and the tuples (provided that the elements within the tuple are StableOrd themselves). I went ahead and just added the impl for these types - is that correct?

Yes, for the two instances you mention that is correct. I'll take a look at other impls shortly.

> Also I'd just like to confirm - is it correct to use unstable sort when calling into_sorted_stable_ord (to_sorted_stable_ord but for UnordSet iterators) with items that are &str? If two &strs are Eq, it should be completely unnoticeable if unstable sort swaps them, correct? Just avoids an allocation ig.

Unstable sort is always correct in two cases:
- When we know that no two elements are equal, e.g. when the sequence to be sorted is the set of keys of a map or set.
- When the items have no observable difference, so that re-ordering them cannot make a difference anywhere. This is the case for `&str` (because we assume that we don't look at verbatim pointer values). Note though, that the entire item has to be considered here: if sorting a sequence of `(&str, SomeOtherType)` with the `&str` as the sort key, then unstable sorting would be allowed to introduce indeterminism.

michaelwoerister replied:
All new impls of `StableOrd` and `ToStableHashKey` look good to me 👍

lcnr commented on compiler/rustc_incremental/src/persist/dirty_clean.rs:
```diff
@@ -197,7 +198,7 @@ impl<'tcx> DirtyCleanVisitor<'tcx> {
         let except = self.except(attr);
         let loaded_from_disk = self.loaded_from_disk(attr);
-        for e in except.iter() {
+        for e in except.to_sorted(&(), false) {
```
`to_sorted_stable_ord` should also work here?

lcnr commented on compiler/rustc_incremental/src/persist/dirty_clean.rs:
```diff
@@ -376,18 +377,18 @@ impl<'tcx> DirtyCleanVisitor<'tcx> {
             };
             self.checked_attrs.insert(attr.id);
-            for label in assertion.clean {
+            assertion.clean.to_sorted(&(), false).iter().for_each(|label| {
```
and here?

afaict all uses of `to_sorted(&(), _)` should be replaceable with `to_sorted_stable_ord` 🤔 please go through all of these calls you're added and replace them

lcnr commented on compiler/rustc_incremental/src/persist/fs.rs:
```diff
@@ -657,62 +658,63 @@ pub fn garbage_collect_session_directories(sess: &Session) -> io::Result<()> {
-        .into_iter()
-        .map(|lock_file_name| {
+    let lock_file_to_session_dir: UnordMap<String, Option<String>> =
+        UnordMap::from(lock_files.into_items().map(|lock_file_name| {
```
can this be `.into()` instead, given that you have an explicit type annotation on the let binding?

lcnr reviewed (commented):
partially went through this PR but would like to see a lot of small changes from `to_sorted` to `to_sorted_stable_ord`. Please resolve this first and then reassign

ndrewxie committed:
Fixed to_sorted => to_sorted_stable_ord

ndrewxie commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

lcnr commented on compiler/rustc_incremental/src/persist/fs.rs:
```diff
@@ -670,11 +672,16 @@ pub fn garbage_collect_session_directories(sess: &Session) -> io::Result<()> {
-    for (lock_file_name, directory_name) in &lock_file_to_session_dir {
+    let lock_file_to_session_dir_iter = lock_file_to_session_dir
+        .items()
+        .map(|(file, dir)| (file.as_str(), dir.as_ref().map(|y| y.as_str())));
```
does `String`/`&String` not implement `StableOrd`? it should

lcnr commented on compiler/rustc_incremental/src/persist/fs.rs:
```diff
@@ -685,34 +692,34 @@ pub fn garbage_collect_session_directories(sess: &Session) -> io::Result<()> {
-        .filter_map(|(lock_file_name, directory_name)| directory_name.map(|n| (lock_file_name, n)))
-        .collect();
+    let lock_file_to_session_dir: UnordMap<String, String> =
+        UnordMap::from(lock_file_to_session_dir.into_items().filter_map(
+            |(lock_file_name, directory_name)| directory_name.map(|n| (lock_file_name, n)),
+        ));
```
```suggestion
        lock_file_to_session_dir.into_items().filter_map(
            |(lock_file_name, directory_name)| directory_name.map(|n| (lock_file_name, n)),
        ).into();
```

lcnr commented on compiler/rustc_incremental/src/persist/fs.rs:
```diff
@@ -721,103 +728,103 @@ pub fn garbage_collect_session_directories(sess: &Session) -> io::Result<()> {
+                );
+            }
+            None
+        });
+    let deletion_candidates = UnordMap::from(deletion_candidates);
```
```suggestion
        }).into();
```

lcnr commented on compiler/rustc_incremental/src/persist/work_product.rs:
```diff
@@ -46,7 +46,12 @@ pub fn copy_cgu_workproduct_to_incr_comp_cache_dir(
-    for (_, path) in &work_product.saved_files {
+    for path in work_product
+        .saved_files
+        .items()
+        .map(|(_, path)| path.as_str())
+        .into_sorted_stable_ord(false)
```
`&String` should implement `StableOrd`

ndrewxie replied:
Hmm, apparently not. Would it be correct to impl StableOrd for `&T` if `T` is StableOrd? I believe it should but just checking in (if not, then I'll just add the impl for `&String` on its own)

lcnr replied:
> Would it be correct to impl StableOrd for &T if T is StableOrd?

yes

lcnr commented on compiler/rustc_interface/src/queries.rs:
```diff
@@ -195,7 +196,8 @@ impl<'tcx> Queries<'tcx> {
                         sess.time("blocked_on_dep_graph_loading", || future.open().open(sess));
-
+                    let prev_work_products =
+                        FxIndexMap::from_iter(prev_work_products.into_sorted(&(), false));
```
```suggestion
                        FxIndexMap::from_iter(prev_work_products.into_sorted_stable_ord(false));
```

lcnr reviewed (commented):
I have to admit that the `bool` argument to `into_sorted_stable_ord` is very confusing to me and even after looking it up in the documentation I always have to wait a second and think about whether it is correct.

I would like us to either add a `const CAN_USE_UNSTABLE_SORT: bool` to `StableOrd` or alternatively always use a stable sort and only add unstable sorting where it is actually performance critical.

Please either do that in this PR or I will open a github issue once your PR lands.

Other thoughts: `UnordItems` using `into()` via `From` instead of having a method called `collect` is also confusing imo.

after dealing with review comments:

r=me,michaelwoerister

ndrewxie committed:
Removed stable/unstable sort arg from into_sorted_stable_ord, fixed a few misc issues, added collect to UnordItems

ndrewxie committed:
fixup! Removed stable/unstable sort arg from into_sorted_stable_ord, fixed a few misc issues, added collect to UnordItems

ndrewxie commented:
Alright, I applied the review suggestions and switched UnordItems::into_sorted_stable_ord to use CAN_USE_UNSTABLE_SORT, and added collect to UnordItems

Also feel free to ignore the clippy comment above, I messed up some submodules stuff but it should be fixed now.

@rustbot label -S-waiting-on-author +S-waiting-on-review

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5207454660/jobs/9395017787) [(plain)](https://github.com/rust-lang/rust/commit/4ae250bf4e285cbd0a474207e8e3fc8d2c5ed9ee/checks/14095238735/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking clippy_lints v0.1.72 (/checkout/src/tools/clippy/clippy_lints)
error[E0061]: this method takes 0 arguments but 1 argument was supplied
   --> src/tools/clippy/clippy_lints/src/wildcard_imports.rs:163:81
    |
163 |                 let mut imports = used_imports.items().map(ToString::to_string).into_sorted_stable_ord(false);
    |                                                                                                        |
    |                                                                                                        unexpected argument of type `bool`
    |                                                                                                        help: remove the extra argument
    |
```

</details>

ndrewxie committed:
Whoops, submodule change was actually valid - undoing fixup

michaelwoerister commented:
Thanks, @ndrewxie -- looks good!

@bors r=lcnr,michaelwoerister

bors commented:
:pushpin: Commit 3f324a8b7d62f44f97ba76eeb76b67d6ff0bf744 has been approved by `lcnr,michaelwoerister`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr,michaelwoerister 3f324a8b7d62f44f97ba76eeb76b67d6ff0bf744 -->
<!-- homu: {"type":"Approved","sha":"3f324a8b7d62f44f97ba76eeb76b67d6ff0bf744","approver":"lcnr,michaelwoerister","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 3f324a8b7d62f44f97ba76eeb76b67d6ff0bf744 with merge a0df04c0f2b9c0415c53e3cee8c4f9fa394a37b2...
<!-- homu: {"type":"BuildStarted","head_sha":"3f324a8b7d62f44f97ba76eeb76b67d6ff0bf744","merge_sha":"a0df04c0f2b9c0415c53e3cee8c4f9fa394a37b2"} -->

lcnr commented on compiler/rustc_incremental/src/persist/dirty_clean.rs:
```diff
@@ -376,15 +377,17 @@ impl<'tcx> DirtyCleanVisitor<'tcx> {
             };
             self.checked_attrs.insert(attr.id);
-            for label in assertion.clean {
+            for label in assertion.clean.items().map(|x| x.as_str()).into_sorted_stable_ord() {
```
```suggestion
            for label in assertion.clean.items().into_sorted_stable_ord() {
```

lcnr commented on compiler/rustc_query_system/src/dep_graph/dep_node.rs:
```diff
@@ -247,3 +247,14 @@ impl<HCX> HashStable<HCX> for WorkProductId {
+}
+unsafe impl StableOrd for WorkProductId {
+    // Fingerprint can use unstable (just a tuple of `u64`s), so WorkProductId can as well
+    const CAN_USE_UNSTABLE_SORT: bool = true;
```
```suggestion
    const CAN_USE_UNSTABLE_SORT: bool = FingerPrint::CAN_USE_UNSTABLE_SORT; 
```
:grin:

lcnr reviewed (commented):
a few nits for code which can be cleaned up now. Feel free to do them in a separate PR once this has landed

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5208583727/jobs/9400977021)
Approved by: lcnr,michaelwoerister
Pushing a0df04c0f2b9c0415c53e3cee8c4f9fa394a37b2 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lcnr,michaelwoerister","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5208583727/jobs/9400977021"},"merge_sha":"a0df04c0f2b9c0415c53e3cee8c4f9fa394a37b2"} -->
